test/hotspot/jtreg/compiler/c2/cr7200264/TestIntVect.java

```diff
@@ -21,622 +21,763 @@
+            System.exit(97);
+        }
+        System.out.println("PASSED");
+
+    }
+
+    int test_sum(int[] a1) {
+        int sum = 0;
+        for (int i = 0; i < a1.length; i+=1) {
+            sum += a1[i];
+        }
+        return sum;
+    }
+
+    @Test
+    @IR(counts = { IRNode.ADD_VI, ">= 1" },
+        applyIfCPUFeature = {"sse2", "true"})
+    void test_addc(int[] a0, int[] a1) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]+VALUE);
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.ADD_VI, ">= 1" },
+        applyIfCPUFeature = {"sse2", "true"})
+    void test_addv(int[] a0, int[] a1, int b) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]+b);
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.ADD_VI, ">= 1" },
+        applyIfCPUFeature = {"sse2", "true"})
+    void test_adda(int[] a0, int[] a1, int[] a2) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]+a2[i]);
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.ADD_VI, ">= 1" },
+        applyIfCPUFeature = {"sse2", "true"})
+    void test_subc(int[] a0, int[] a1) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]-VALUE);
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.SUB_VI, ">= 1" },
+        applyIfCPUFeature = {"sse2", "true"})
+    void test_subv(int[] a0, int[] a1, int b) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]-b);
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.SUB_VI, ">= 1", },
+        applyIfCPUFeature = {"sse2", "true"})
+    void test_suba(int[] a0, int[] a1, int[] a2) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]-a2[i]);
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.SUB_VI, ">= 1", IRNode.LSHIFT_VI, ">= 1" },
+        applyIfCPUFeature = {"sse2", "true"})
+    void test_mulc(int[] a0, int[] a1) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]*VALUE);
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.SUB_VI, ">= 1", IRNode.LSHIFT_VI, ">= 1" },
+        applyIfCPUFeature = {"sse2", "true"})
+    void test_mulc_n(int[] a0, int[] a1) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]*(-VALUE));
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.MUL_VI, ">= 1" },
+        applyIfCPUFeature = {"sse4.1", "true"})
+    void test_mulv(int[] a0, int[] a1, int b) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]*b);
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.MUL_VI, ">= 1" },
+        applyIfCPUFeature = {"sse4.1", "true"})
+    void test_mula(int[] a0, int[] a1, int[] a2) {
+        for (int i = 0; i < a0.length; i+=1) {
+            a0[i] = (int)(a1[i]*a2[i]);
+        }
+    }
+
+    void test_divc(int[] a0, int[] a1) {
```

Comment:
I was curious about that and it actually does:
```
TraceNewVectors [SuperWord]:  744  LoadVector  === 380 693 677  [[ 675 671 669 664 660 658 558 554 552 145 150 154 ]]  @int[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; mismatched #vectorx[4]:{int} !orig=[676],[559],[134] !jvms: Test::test_divc @ bci:12 (line 36)
TraceNewVectors [SuperWord]:  746  RShiftVI  === _ 744 745  [[ 668 657 551 155 ]]  #vectorx[4]:{int} !orig=[669],[552],[154] !jvms: Test::test_divc @ bci:15 (line 36)
TraceNewVectors [SuperWord]:  747  VectorCastI2X  === _ 744  [[ 674 663 557 146 ]]  #vectory[4]:{long} !orig=[675],[558],[145] !jvms: Test::test_divc @ bci:15 (line 36)
TraceNewVectors [SuperWord]:  748  Replicate  === _ 144  [[ ]]  #vectory[4]:{long}
TraceNewVectors [SuperWord]:  749  MulVL  === _ 747 748  [[ 673 662 556 148 ]]  #vectory[4]:{long} !orig=[674],[557],[146] !jvms: Test::test_divc @ bci:15 (line 36)
TraceNewVectors [SuperWord]:  751  RShiftVL  === _ 749 750  [[ 672 661 555 149 ]]  #vectory[4]:{long} !orig=[673],[556],[148] !jvms: Test::test_divc @ bci:15 (line 36)
TraceNewVectors [SuperWord]:  752  VectorCastL2X  === _ 751  [[ 671 660 554 150 ]]  #vectorx[4]:{int} !orig=[672],[555],[149] !jvms: Test::test_divc @ bci:15 (line 36)
TraceNewVectors [SuperWord]:  753  AddVI  === _ 752 744  [[ 670 659 553 152 ]]  #vectorx[4]:{int} !orig=[671],[554],[150] !jvms: Test::test_divc @ bci:15 (line 36)
TraceNewVectors [SuperWord]:  755  RShiftVI  === _ 753 754  [[ 668 657 551 155 ]]  #vectorx[4]:{int} !orig=[670],[553],[152] !jvms: Test::test_divc @ bci:15 (line 36)
TraceNewVectors [SuperWord]:  756  SubVI  === _ 755 746  [[ 666 656 549 176 ]]  #vectorx[4]:{int} !orig=[668],[551],[155] !jvms: Test::test_divc @ bci:15 (line 36)
TraceNewVectors [SuperWord]:  757  StoreVector  === 687 693 667 756  [[ 374 693 372 179 ]]  @int[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; mismatched  Memory: @int[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; !orig=[666],[549],[176],575 !jvms: Test::test_divc @ bci:16 (line 36)
```
I have not checked any other methods but it might indeed be possible to vectorize some them. I think it's a good idea to check all methods and add a comment with a short explanation why it's not possible. Thanks @eme64 for helping out to do that.

Reply:
Should we also add checks for these vectors (same for `test_divc_n()`)?

Reply:
@chhagedorn: Do you mean that `test_divc` and `test_divc_n` vectorize after JDK-8282365? They don't vectorize on my machine (on this PR).

Reply:
I just checked on my machine (on top of commit fb822e49f2a84423c8fd17db2e95bbdd5e7ec191) and these division tests do seem to vectorize, this is e.g. the innermost loop in `test_divc` right before code emission:

![test_divc](https://github.com/openjdk/jdk/assets/8792647/129d51c2-a1ad-4d02-ab81-02cd849af36f)

Here are my processor features in case it helps (subset of `lscpu` output):

```
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         39 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  12
  On-line CPU(s) list:   0-11
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz
    CPU family:          6
    Model:               158
    Thread(s) per core:  2
    Core(s) per socket:  6
(...)
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mc
                         a cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss 
                         ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art
                          arch_perfmon pebs bts rep_good nopl xtopology nonstop_
                         tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cp
                         l vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid ss
                         e4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes 
                         xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_f
                         ault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhan
                         ced tpr_shadow flexpriority ept vpid ept_ad fsgsbase ts
                         c_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed ad
                         x smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsav
                         es dtherm ida arat pln pts hwp hwp_notify hwp_act_windo
                         w hwp_epp vnmi md_clear flush_l1d arch_capabilities
(...)

```

Reply:
> @chhagedorn: Do you mean that `test_divc` and `test_divc_n` vectorize after JDK-8282365? They don't vectorize on my machine (on this PR).

Ah, I see. Yes, as for Roberto, it does vectorize on my machine as well.

Reply:
Thanks for the clarification @robcasloz and @chhagedorn. I've investigated now, and they do vectorize on my machine as well. I was confused because, before the change below, the IR framework did not register the nodes (wrong vector size of 4 instead of the default of 8). Is that expected, and should we specify something else instead of the catch-all `IRNode.VECTOR_SIZE_ANY`?

```
     @Test
-    @IR(counts = { IRNode.ADD_VI,    "> 0",
-                   IRNode.RSHIFT_VI, "> 0",
-                   IRNode.SUB_VI,    "> 0" },
+    @IR(counts = { IRNode.ADD_VI,    IRNode.VECTOR_SIZE_ANY, "> 0",
+                   IRNode.RSHIFT_VI, IRNode.VECTOR_SIZE_ANY, "> 0",
+                   IRNode.SUB_VI,    IRNode.VECTOR_SIZE_ANY, "> 0" },
         applyIfCPUFeatureOr = {"sse2", "true", "asimd", "true"})
     void test_divc(int[] a0, int[] a1) {
         for (int i = 0; i < a0.length; i+=1) {
@@ -519,9 +519,9 @@ void test_divc(int[] a0, int[] a1) {
     }
 
     @Test
-    @IR(counts = { IRNode.ADD_VI,    "> 0",
-                   IRNode.RSHIFT_VI, "> 0",
-                   IRNode.SUB_VI,    "> 0" },
+    @IR(counts = { IRNode.ADD_VI,    IRNode.VECTOR_SIZE_ANY, "> 0",
+                   IRNode.RSHIFT_VI, IRNode.VECTOR_SIZE_ANY, "> 0",
+                   IRNode.SUB_VI,    IRNode.VECTOR_SIZE_ANY, "> 0" },
         applyIfCPUFeatureOr = {"sse2", "true", "asimd", "true"})
     void test_divc_n(int[] a0, int[] a1) {
         for (int i = 0; i < a0.length; i+=1) {
```

Reply:
@dlunde do you understand what factors determine the length of the vector? Why is the default of `IRNode.VECTOR_SIZE_MAX` not working?

Reply:
> @dlunde do you understand what factors determine the length of the vector? Why is the default of IRNode.VECTOR_SIZE_MAX not working?

Perhaps C2 hits the loop unrolling limit? @dlunde you can test this by trying out a large value for `-XX:LoopUnrollLimit`. But even if this turned out to be the case, I would still suggest using `IRNode.VECTOR_SIZE_ANY` rather than forcing a higher loop unroll limit value for the tests.

Reply:
Ah, I see what is the issue here: the loop does not just contain `int` vectors but also `long` vectors.
Specifically, I see a `VectorCastI2X` and `VectorCastL2X` nodes in the loop, which converst `int` to/from `long`.
Hence, if you have a `32 byte` vector, you can only have `4 long`, and so the loop-unrolling is limited to 4x.
And then you only see `4 int` vectors, when you were expecting `8 int` vectors.

You will probably be able to fix the issue with this:
`IRNode.VECTOR_SIZE + "min(max_int, max_long)"`

For more examples, check out:
`grep "IRNode.VECTOR_SIZE +" test/hotspot/jtreg/ -r`

Reply:
@eme64: Thanks, I'll check that `IRNode.VECTOR_SIZE + "min(max_int, max_long)"` works and use that.

@robcasloz: Thanks for the good idea; it indeed had to do with an unrolling limit. Just to clarify regarding `IRNode.VECTOR_SIZE_ANY` (after discussions with @eme64): I also thought `IRNode.VECTOR_SIZE_ANY` seemed like a good solution, but learned that `IRNode.VECTOR_SIZE_MAX` is the default because less than max vectorization usually indicates a problem. I may be wrong, but I believe `IRNode.VECTOR_SIZE_MAX` also varies depending on the platform. Therefore, I'll go with what Emanuel suggests for the `test_divc` and `test_divc_n` IR checks.